### PR TITLE
feat: Add ca to check the validation of the member clusters' server certificate

### PR DESCRIPTION
**Pull request by jwcesign**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/4182

**Special notes for your reviewer**:
If I change the following line to false:
https://github.com/karmada-io/karmada/blob/3267155766f69083d211c4bc804e8bda4b18fd4d/pkg/util/proxy/proxy.go#L79

It will have the following error in `karmada-aggregated-apiserver`:
01d9281bfb18dfd30d78ab197884c43

And karmadactl couldn't list the pods in member clusters:
d2a7ad79b3a939f7263191265be2176

With this PR, everything works fine

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-aggregated-apiserver`: Add ca to check the validation of the member clusters' server certificate.
```

### CODE REVIEW

**Review by zishen — COMMENTED** [review]
Good for validating  the member clusters' server certificate

**Comment by RainbowMango** [comment]
The unit test is failing, @jwcesign :
https://github.com/karmada-io/karmada/actions/runs/6651924401/job/18075272929?pr=4183#step:4:5247

File: pkg/util/proxy/proxy.go [hunk 1]
```diff
@@ -94,14 +102,39 @@ func NewThrottledUpgradeAwareProxyHandler(location *url.URL, transport http.Roun
 	return proxy.NewUpgradeAwareHandler(location, transport, wrapTransport, upgradeRequired, proxy.NewErrorResponder(responder))
 }
 
+func GetTlsConfigForCluster(ctx context.Context, cluster *clusterapis.Cluster, secretGetter SecretGetterFunc) (*tls.Config, error) {
+	if cluster.Spec.SecretRef == nil {
```
Comment (RainbowMango) [inline pkg/util/proxy/proxy.go:106]: Can you remind me when the secret would not exist?

File: pkg/util/proxy/proxy.go [hunk 2]
```diff
@@ -94,14 +102,39 @@ func NewThrottledUpgradeAwareProxyHandler(location *url.URL, transport http.Roun
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get CA bundle for cluster %s: %v", cluster.Name, err)
+	}
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM([]byte(caBundle))
+	return &tls.Config{
+		RootCAs:            caCertPool,
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: cluster.Spec.InsecureSkipTLSVerification, //nolint:gosec
```
Comment (RainbowMango) [inline pkg/util/proxy/proxy.go:126]: Can we remove the `nolint:gosec` now?

Same as above.

File: pkg/util/proxy/proxy.go (hunk 1 above)
Comment (jwcesign) [reply pkg/util/proxy/proxy.go:106]: From my point of view, it couldn't be empty. But I saw it's an optional field in the struct comment.

File: pkg/util/proxy/proxy.go (hunk 2 above)
Comment (jwcesign) [reply pkg/util/proxy/proxy.go:126]: If delete it, there is an error: `cluster.Spec.InsecureSkipTLSVerification may be true`.

But it's allowed to be true(if users want to do it).

File: pkg/util/proxy/proxy.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/util/proxy/proxy.go:106]: For `Pull` mode cluster, the `karmada-agent` may not set the secret, see the usage from `karmada-agent`:
```bash
      --report-secrets strings                                                                                                                                                                     
                The secrets that are allowed to be reported to the Karmada control plane during registering. Valid values are 'KubeCredentials', 'KubeImpersonator' and 'None'. e.g
                'KubeCredentials,KubeImpersonator' or 'None'. (default [KubeCredentials,KubeImpersonator])

```

File: pkg/util/proxy/proxy.go (hunk 2 above)
Comment (RainbowMango) [reply pkg/util/proxy/proxy.go:126]: Before:
```
pkg/util/proxy/proxy.go:109:24: G402: TLS InsecureSkipVerify set true. (gosec)
			InsecureSkipVerify: true,
```
After:
```
pkg/util/proxy/proxy.go:109:24: G402: TLS InsecureSkipVerify may be true. (gosec)
			InsecureSkipVerify: cluster.Spec.InsecureSkipTLSVerification,
```
we might need a comment to explain it.

File: pkg/util/proxy/proxy.go [hunk 3]
```diff
@@ -94,14 +102,42 @@ func NewThrottledUpgradeAwareProxyHandler(location *url.URL, transport http.Roun
 	return proxy.NewUpgradeAwareHandler(location, transport, wrapTransport, upgradeRequired, proxy.NewErrorResponder(responder))
 }
 
+func GetTlsConfigForCluster(ctx context.Context, cluster *clusterapis.Cluster, secretGetter SecretGetterFunc) (*tls.Config, error) {
+	if cluster.Spec.SecretRef == nil {
+		return &tls.Config{
+			MinVersion: tls.VersionTLS13,
+			// 1.If no comment 'nolint:gosec', there will be golint error 'TLS InsecureSkipVerify may be true'.
+			// 2.When the cluster CustomResource is not configured with CA(For Pull mode cluster, the karmada-agent may not set the CA),
+			//   InsecureSkipTLSVerification of cluster CustomResource must be true to ensure karmada-aggregated-apiserver working fine.
```
Comment (RainbowMango) [inline pkg/util/proxy/proxy.go:111]:
```suggestion

			// Ignore false positive warning: "TLS InsecureSkipVerify may be true. (gosec)"
			// Whether to skip server certificate verification depends on the
			// configuration(.spec.insecureSkipTLSVerification, defaults to false) in a Cluster object.
```

File: pkg/util/proxy/proxy.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/util/proxy/proxy.go:106]:
```suggestion
	// The secret is optional for a pull-mode cluster, if no secret just returns a config with root CA unset.
	if cluster.Spec.SecretRef == nil {
```
Maybe we can have a comment here.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**